### ISSUE TITLE
modernize-use-override: ignore when `virtual` is from a macro 

### ISSUE THREAD

**Issue by saschanaz** [llvm/llvm-project#53949]
[Mozilla uses `NS_IMETHOD` macro a lot](https://searchfox.org/mozilla-central/search?q=NS_IMETHOD&path=) to implement a certain kind of virtual methods.

```cpp
#ifdef XP_WIN
#  define NS_IMETHOD_(type) virtual type __stdcall
#else
#  define NS_IMETHOD_(type) virtual type
#endif

#define NS_IMETHOD NS_IMETHOD_(nsresult)
```

We use this to implement both a base method and overrides:

```cpp
class Foo {
  NS_IMETHOD Foo();
};

class Bar: public Foo {
  NS_IMETHOD Foo() override;
};
```

`modernize-use-override` whines at this kind of code:

```
'virtual' is redundant since the function is already declared 'override' [modernize-use-override]
```

It's redundant, yes, but that's okay as it's hidden behind the macro. I think the linter should ignore such case.

**Comment by njames93** [comment]
I'm not sure this is really an use-case that should be supported, take this:
```c++
#define VIRTUAL virtual

struct Foo {
  VIRTUAL int getID() const;
};

struct Bar : Foo {
  VIRTUAL int getID() override;
};
```
In this case I'd argue we definitely should have a warning.
Maybe there could be merit in your case when the macro isn't just for defining the virtual keyword

**Comment by saschanaz** [comment]
Yes, it's more than just a single-keyword macro. But even it was a single-keyword one, I think something like this could still make a sense:

```cpp
#define IMETHODDECL virtual
struct Foo {
  IMETHODDECL void foo();
};
struct Bar : Foo {
  IMETHODDECL void foo() override; // it's probably there for a reason, maybe for some style guide.
};
```

**Comment by emilio** [comment]
It'd be somewhat weird, I think, to have a macro that expands only to the virtual keyword, right?

There are a few other warnings that are silenced as a result of macro expansion, iirc (though on the phone, can try to look for precedents later).

**Comment by saschanaz** [comment]
Haven't checked yet, but I see quite some amount of `isMacroID()` checks in clang-tidy.

**Comment by njames93** [comment]
> Haven't checked yet, but I see quite some amount of `isMacroID()` checks in clang-tidy.

Those are often there as FixIts in macro code often causes compilation issues

**Comment by saschanaz** [comment]
I see such warn-but-no-fix cases (just like `modernize-use-override`) but also no-warn-at-all ones:

https://github.com/llvm/llvm-project/blob/4a097efe7784767b7d12ffcb8f2b22b9f4d045e2/clang-tools-extra/clang-tidy/cppcoreguidelines/InitVariablesCheck.cpp#L62-L78

**Comment by AaronBallman** [comment]
We often will suppress diagnostics in the presence of macros (same as fix-its) because macros make it very difficult to know whether the user has the ability to silence the diagnostic themselves or not due to configuration differences. I think this is one of those cases -- Mozilla uses this in a number of places, and that's indicative of a code pattern that exists in the wild where this otherwise useful check has to be disabled. That may be worth considering adding a configuration option over it. That said, I definitely think it's a questionable to hide the `virtual` behind a macro as that drastically impacts the intent of the interface when the macro is disabled.

**Comment by vbvictor** [comment]
https://godbolt.org/z/TMfqzfE7e
